File: Renaissance.AspNet/frontend/Renaissance.Web/Services/AuthStateService.cs
using Microsoft.AspNetCore.Components.Server.ProtectedBrowserStorage;
using Renaissance.Web.Models;

namespace Renaissance.Web.Services;

public class AuthStateService
{
    public const string TokenStorageKey = "renaissance.auth.token";

    private readonly ProtectedSessionStorage _session;

    public AuthStateService(ProtectedSessionStorage session)
    {
        _session = session;
    }

    public bool IsReady { get; private set; }
    public string? Token { get; private set; }
    public CurrentUserDto? User { get; private set; }
    public bool IsAuthenticated => !string.IsNullOrWhiteSpace(Token) && User is not null;

    public event Action? Changed;

    public bool HasModule(AppModule module)
        => User?.IsAdministrator == true || (User?.Modules.Contains(module) ?? false);

    public async Task<string?> ReadStoredTokenAsync()
    {
        try
        {
            var result = await _session.GetAsync<string>(TokenStorageKey);
            return result.Success ? result.Value : null;
        }
        catch
        {
            return null;
        }
    }

    public async Task SignInAsync(string token, CurrentUserDto user)
    {
        Token = token;
        User = user;
        IsReady = true;
        await _session.SetAsync(TokenStorageKey, token);
        Changed?.Invoke();
    }

    public void SetSession(string token, CurrentUserDto? user)
    {
        Token = token;
        User = user;
        IsReady = true;
        Changed?.Invoke();
    }

    public void MarkReady()
    {
        IsReady = true;
        Changed?.Invoke();
    }

    public async Task SignOutAsync()
    {
        Token = null;
        User = null;
        IsReady = true;
        try
        {
            await _session.DeleteAsync(TokenStorageKey);
        }
        catch
        {
            // Ignore storage failures on logout.
        }

        Changed?.Invoke();
    }
}

File: Renaissance.AspNet/frontend/Renaissance.Angular/src/app/core/services/auth-state.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { AppModule, CurrentUserDto } from '../models/auth.models';

const TOKEN_KEY = 'renaissance.auth.token';

@Injectable({ providedIn: 'root' })
export class AuthStateService {
    private readonly _ready$ = new BehaviorSubject<boolean>(false);
    private readonly _user$ = new BehaviorSubject<CurrentUserDto | null>(null);
    private _token: string | null = null;

    readonly ready$ = this._ready$.asObservable();
    readonly user$ = this._user$.asObservable();

    get token(): string | null {
        return this._token;
    }

    get user(): CurrentUserDto | null {
        return this._user$.value;
    }

    get isAuthenticated(): boolean {
        return !!this._token && !!this.user;
    }

    get isReady(): boolean {
        return this._ready$.value;
    }

    hasModule(module: AppModule): boolean {
        const user = this.user;
        return !!user && (user.isAdministrator || user.modules.includes(module));
    }

    restoreFromStorage(): void {
        this._token = sessionStorage.getItem(TOKEN_KEY);
        this._ready$.next(true);
    }

    signIn(token: string, user: CurrentUserDto): void {
        this._token = token;
        this._user$.next(user);
        sessionStorage.setItem(TOKEN_KEY, token);
        this._ready$.next(true);
    }

    setUser(user: CurrentUserDto | null): void {
        this._user$.next(user);
    }

    signOut(): void {
        this._token = null;
        this._user$.next(null);
        sessionStorage.removeItem(TOKEN_KEY);
        this._ready$.next(true);
    }
}
